[CSClosure] Add support for projected/wrapper values

xedin opened the pull request:
To support referencing projected and/or wrapped var
in a closure solver needs to lookup a type of their
originator and based on the wrapper type compute and
assign types to projection and/or wrapper.

Resolves: https://github.com/apple/swift/issues/59295
Resolves: rdar://94506352

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSClosure] Add support for projected/wrapper values

To support referencing projected and/or wrapped var
in a closure solver needs to lookup a type of their
originator and based on the wrapper type compute and
assign types to projection and/or wrapper.

Resolves: https://github.com/apple/swift/issues/59295
Resolves: rdar://94506352

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

hborla commented on lib/Sema/CSClosure.cpp:
```diff
@@ -96,6 +96,29 @@ class TypeVariableRefFinder : public ASTWalker {
 
+      if (auto *wrappedVar = var->getOriginalWrappedProperty()) {
+        auto outermostWrapperAttr =
+            wrappedVar->getAttachedPropertyWrappers().back();
```
Hmm, `getAttachedPropertyWrappers().front()` should give the outermost wrapper attribute

xedin replied:
Indeed, I thought that it was in reverse, going to add a method to make it harder to get wrong...

xedin replied:
Fixed!

hborla commented on validation-test/Sema/type_checker_crashers_fixed/issue59294.swift:
```diff
@@ -25,18 +26,94 @@ struct WrapperValue<Value> {
+        @OuterWrapper @WrapperValue var value = $0
+        if true {
+          let _: Bool = _value == $0
+          let _: OuterWrapper<Bool> = $value
```
The type of the projected value here should be `OuterWrapper<WrapperValue<Bool>>`

xedin replied:
Fixed by using correct "outer" wrapper, thank you!

xedin committed:
[TypeChecker] NFC: Add a dedicated method to get outermost attached wrapper

The outermost wrapper is the one at index `0` in the wrapper list
but it's easy for humans to make a reverse assumption since outermost
is the back of the list. Let's add a dedicated method to reduce error
probability of the property wrapper APIs.

xedin commented:
@swift-ci please test
